File: src/SSVMtoolbox/r/KGaussian.R
#-------------------------------------------------------------------------#
#Building kernel data matrix, full or reduced.

#Inputs                                                                  
#A: full data set                                                        
#tilde A: can be full or reduced set                                     
#gamma: width parameter; kernel value: exp(-gamma(Ai-Aj)^2)              
                                                                       
#Outputs                                                                 
#K: kernel data using Gaussian kernel                                    
#-------------------------------------------------------------------------#

KGaussian = function (gamma, A, tildeA){
  options(warn=-1)
  row=dim(A)[1]
  #use full kernel
  if (missing('tildeA')){
    AA=matrix(rowSums(A^2), row, row)
    K=exp((-AA-t(AA)+2*A%*%t(A))*gamma)
  } else{ 
    #use reduced kernel
    tilderow=dim(tildeA)[1]
    AA=matrix(rowSums(A^2), row, tilderow)
    tildeAA=matrix(rowSums(tildeA^2), tilderow, row)
    K=exp((-AA-t(tildeAA)+2*A%*%t(tildeA))*gamma)
  }
  return(K)
}

File: src/SSVMtoolbox/matlab/KGaussian.m
function K = KGaussian(gamma, A, tildeA)
%=========================================================================%
% Building kernel data matrix, full or reduced.                           %
%                                                                         %
% Inputs                                                                  %
% A: full data set                                                        %
% tilde A: can be full or reduced set                                     %
% gamma: width parameter; kernel value: exp(-gamma(Ai-Aj)^2)              %
%                                                                         %
% Outputs                                                                 %
% K: kernel data using Gaussian kernel                                    %
% by Kuang-Yao Lee                                                        %
%=========================================================================%

row=size(A,1);
if nargin<3, % square full kernel     
    AA = repmat(sum(A.^2,2),1,row);
    K = exp((-AA-AA'+2*A*A')*gamma);
else   %reduced kernel, or kernel matrix for test data
    AA = repmat(sum(A.^2,2),1,size(tildeA,1));
    tildeAA = repmat(sum(tildeA.^2,2),1,row);
    K = exp((-AA-tildeAA'+2*A*tildeA')*gamma);
end